jobs/blog_validation_jobs.py

```diff
@@ -0,0 +1,159 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Validation Jobs for blog models"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.platform import models
+from jobs import base_jobs
+from jobs import blog_validation_errors
+from jobs.io import ndb_io
+
+import apache_beam as beam
+(blog_models,) = models.Registry.import_models([models.NAMES.blog, ])
+
+
+class BlogPostTitleUniquenessJob(base_jobs.JobBase):
+    """Validates that all the Blog Posts have unique title."""
+
+    def run(self):
+        blog_model_pcoll = (
+            self.pipeline
+            | 'Get every Blog Model' >> (
+                ndb_io.GetModels(
+                    blog_models.BlogPostModel.query(), self.datastoreio_stub
+                ))
+        )
+
+        blogs_with_duplicate_titles = (
+            blog_model_pcoll
+            | 'Generate (title, model) key value pairs' >> (
+                beam.WithKeys(lambda blog_model: blog_model.title)) # pylint: disable=no-value-for-parameter
+            | 'Group pairs by their title' >> beam.GroupByKey()
+            | 'Discard keys' >> beam.Values()
+            | 'Discard models with unique titles' >> (
+                beam.Filter(lambda models: len(models) > 1))
+        )
+
+        return (
+            blogs_with_duplicate_titles
+            | 'Flatten models into a list of errors' >> beam.FlatMap(
+                lambda models: [
+                    blog_validation_errors.DuplicateBlogTitleError(model)
+                    for model in models if model.title != ''
+                ])
+        )
+
+
+class BlogPostUrlUniquenessJob(base_jobs.JobBase):
+    """Validates that all the Blog Posts have unique url."""
+
+    def run(self):
+        blog_model_pcoll = (
+            self.pipeline
+            | 'Get every Blog Post Model' >> (
+                ndb_io.GetModels(
+                    blog_models.BlogPostModel.query(), self.datastoreio_stub
+                ))
+        )
+
+        blogs_with_duplicate_url_fragment = (
+            blog_model_pcoll
+            | 'Generate (url, model) key value pairs' >> (
+                beam.WithKeys(lambda blog_model: blog_model.url_fragment)) # pylint: disable=no-value-for-parameter
+            | 'Group pairs by their url fragment' >> beam.GroupByKey()
+            | 'Discard keys' >> beam.Values()
+            | 'Discard models with unique urls' >> (
+                beam.Filter(lambda models: len(models) > 1))
+        )
```

Comment:
Looks like this is a lot of duplicated code, please try writing a helper PTransform that takes in the `WithKeys` function instead:

```python
class GetModelsWithDuplicatePropertyValues(beam.PTransform):

    def __init__(self, property_name):
        super(GetModelsWithDuplicatePropertyValues, self).__init__()
        self._property_name = property_name

    def expand(self, blog_model_pcoll):
        return (
            blog_model_pcoll
            | 'Generate (%s, model) key value pairs' % self._property_name >> (
                beam.WithKeys(self.get_property_value))
            | 'Group pairs by their %s' % self._property_name >> beam.GroupByKey()
            | 'Discard %s key' % self._property_name >> beam.Values()
            | 'Discard models with unique %s' % self._property_name >> (
                beam.Filter(lambda models: len(models) > 1))
        )

    def get_property_value(self, model):
        return job_utils.get_model_property(model, self._property_name)

# ...

    blogs_with_duplicate_url_fragment = (
        blog_model_pcoll | GetModelsWithDuplicatePropertyValues('url_fragment')
    )
```

Reply:
Done. Thanks